Distance_3/include/CGAL/squared_distance_3_1.h

```diff
@@ -174,6 +174,125 @@ squared_distance(
+    typedef typename K::Vector_3 Vector_3;
+    typedef typename K::RT RT;
+    typedef typename K::FT FT;
+    FT d1=FT(0), d2=FT(0);
+    RT e1 = RT(1), e2 = RT(1);
+    // assert that the segment is valid (non zero length).
+    {
+      Vector_3 diff = construct_vector(seg1.source(), pt);
+      Vector_3 segvec = construct_vector(seg1.source(), seg1.target());
+      RT d = wdot(diff,segvec, k);
+      if (d <= (RT)0){
+        d1 = (FT(diff*diff));
+      }else{
+        RT e = wdot(segvec,segvec, k);
+        if (d > e){
+          d1 = squared_distance(pt, seg1.target(), k);
+        } else{
+          Vector_3 wcr = wcross(segvec, diff, k);
+          d1 = FT(wcr*wcr);
+          e1 = e;
+        }
+      }
+    }
+    {
+      Vector_3 diff = construct_vector(seg2.source(), pt);
+      Vector_3 segvec = construct_vector(seg2.source(), seg2.target());
+      RT d = wdot(diff,segvec, k);
+      if (d <= (RT)0){
+        d2 = (FT(diff*diff));
+      }else{
+        RT e = wdot(segvec,segvec, k);
+        if (d > e){
+          d2 = squared_distance(pt, seg2.target(), k);
+        } else{
+          Vector_3 wcr = wcross(segvec, diff, k);
+          d2 = FT(wcr*wcr);
+          e2 = e;
+        }
+      }
+    }
+    return compare(d1*e2, d2*e1);
+}
+
+template <class K>
+typename K::Comparison_result
+compare_distance(
+    const typename K::Point_3 &pt,
+    const typename K::Point_3 &pt2,
+    const typename K::Segment_3 &seg,
+    const K& k,
+    const Cartesian_tag&)
+{
+    typename K::Construct_vector_3 construct_vector;
+    typedef typename K::Vector_3 Vector_3;
+    typedef typename K::RT RT;
+    typedef typename K::FT FT;
+    RT e2 = RT(1);
+    // assert that the segment is valid (non zero length).
+    FT d1 = squared_distance(pt, pt2, k);
+    FT d2 = FT(0);
+    {
+      Vector_3 diff = construct_vector(seg.source(), pt);
+      Vector_3 segvec = construct_vector(seg.source(), seg.target());
+      RT d = wdot(diff,segvec, k);
+      if (d <= (RT)0){
+        d2 = (FT(diff*diff));
+      }else{
+        RT e = wdot(segvec,segvec, k);
+        if (d > e){
+          d2 = squared_distance(pt, seg.target(), k);
+        } else{
+          Vector_3 wcr = wcross(segvec, diff, k);
+          d2 = FT(wcr*wcr);
+          e2 = e;
+        }
+      }
+    }
+    return compare(d1*e2, d2);
+}
+
+template <class K>
+inline
+typename K::Comparison_result
+compare_distance(
+    const typename K::Point_3 & pt,
+    const typename K::Point_3 & pt2,
+    const typename K::Segment_3 & seg2,
+    const K& k)
+{
+  typedef typename K::Kernel_tag Tag;
+  Tag tag;
+  return compare_distance(pt, pt2, seg2, k, tag);
+}
+
+
+template <class K>
+inline
+typename K::Comparison_result
+compare_distance(
+    const typename K::Point_3 & pt,
+    const typename K::Segment_3 & seg1,
+    const typename K::Segment_3 & seg2,
+    const K& k)
+{
+  typedef typename K::Kernel_tag Tag;
+  Tag tag;
+  return compare_distance(pt, seg1, seg2, k, tag);
+}
```

Comment:
What is the expected match when `tag` is not of type `Cartesian_tag`?

Have you tested the testsuite of `Kernel_23` and `Distance_3`? I have the impression that it will fail to compile with a non-Cartesian kernel.